### PR TITLE
Saving checked VOIs (438)

### LINKED ISSUE #438: Selected contours don't save
Selected contours don't save when changing item in patient tree (left menu)

### PR DESCRIPTION
**Pull request by rozchlastywacz**
Fixed saving checked VOI when changing patients by:
- adding new object that holds gui state
- adding new method in ListWidget, which checks passed items
- changing set_patient method to properly use new state object

closes #438 
- #438

### CODE REVIEW

File: pytripgui/canvas_vc/canvas_controller.py [hunk 1]
```diff
@@ -80,19 +85,35 @@ def _update_canvas_view(self):
...
                 self._ui.plot_let(self._model.let)
 
-    def set_patient(self, patient, state):
+    def set_patient(self, patient, state: PatientGuiState):
         self._ui.clear()
-        if state is None:
-            state = ProjectionSelector()
-        self._model = PlotModel(state)
+
+        if state and state.projection_selector:
```
Comment (grzanka) [inline pytripgui/canvas_vc/canvas_controller.py:91]: explain this conditional blocks in more details, i.e. what situation projection_selector is set and in which - not

File: pytripgui/canvas_vc/gui_state.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from typing import Optional, List
+
+from pytripgui.canvas_vc.projection_selector import ProjectionSelector
+
+
+class PatientGuiState:
+    """
+    This class holds information about viewing patient
+        - on which slice user stopped
```
Comment (grzanka) [inline pytripgui/canvas_vc/gui_state.py:9]: on which slice which action was stopped ?

File: pytripgui/view/qt_view_adapter.py [hunk 3]
```diff
@@ -172,6 +172,14 @@ def checked_items(self):
                 selected.append(widget.data(Qt.UserRole))
         return selected
 
+    def check_items(self, items, lambda_names):
+        to_be_checked = [
```
Comment (grzanka) [inline pytripgui/view/qt_view_adapter.py:176]: to be checked sounds like be checked by some external tool, maybe we use word selected ? or checkbox selected ?

File: pytripgui/canvas_vc/canvas_controller.py (hunk 1 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/canvas_controller.py:91]: Default value is None, so I check if it is in that condition, but I will check if it is possible for it to be None in that method

File: pytripgui/view/qt_view_adapter.py [hunk 4]
```diff
@@ -172,6 +172,14 @@ def checked_items(self):
...
+            list_item for list_item in self._items
+            if lambda_names(list_item.data(Qt.UserRole)) in [lambda_names(item) for item in items]
```
Comment (grzanka) [inline pytripgui/view/qt_view_adapter.py:178]: what is lambda_names ?

**Review by grzanka — CHANGES_REQUESTED** [review]

File: pytripgui/view/qt_view_adapter.py (hunk 4 above)
Comment (rozchlastywacz) [reply pytripgui/view/qt_view_adapter.py:178]: copied from another method signature - lambda that gets name from passed items

File: pytripgui/view/qt_view_adapter.py (hunk 3 above)
Comment (rozchlastywacz) [reply pytripgui/view/qt_view_adapter.py:176]: or maybe `items_to_check`?

Comment (rozchlastywacz) [reply pytripgui/view/qt_view_adapter.py:176]: you can select item without checking it, so those words are not interchangeable in that case

File: pytripgui/canvas_vc/gui_state.py (hunk 2 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/gui_state.py:9]: you mean that `scrolling` is missing?
`on which slice user stopped scrolling`
even better
`on which slice user stopped scrolling on each plane`

File: pytripgui/canvas_vc/canvas_controller.py (hunk 1 above)
Comment (grzanka) [reply pytripgui/canvas_vc/canvas_controller.py:91]: I was more concerned why check for `projection_selector` at all in `set_patient` method.
It is not obvious from the code, hence should go into the comment

File: pytripgui/canvas_vc/gui_state.py (hunk 2 above)
Comment (grzanka) [reply pytripgui/canvas_vc/gui_state.py:9]: last one is best

File: pytripgui/view/qt_view_adapter.py (hunk 4 above)
Comment (grzanka) [reply pytripgui/view/qt_view_adapter.py:178]: it is not unclear, what about `name_extractor` ? Are we forced to use lambdas here, I suspect function will work as well

File: pytripgui/view/qt_view_adapter.py (hunk 3 above)
Comment (grzanka) [reply pytripgui/view/qt_view_adapter.py:176]: we should be explicit then, like `toogle_checkbox_for_items` or `set_checkbox_enabled` or  `set_tickbox` ?

File: pytripgui/canvas_vc/canvas_controller.py (hunk 1 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/canvas_controller.py:91]: Done

File: pytripgui/canvas_vc/gui_state.py (hunk 2 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/gui_state.py:9]: Done

File: pytripgui/view/qt_view_adapter.py (hunk 3 above)
Comment (rozchlastywacz) [reply pytripgui/view/qt_view_adapter.py:176]: Done

File: pytripgui/view/qt_view_adapter.py (hunk 4 above)
Comment (rozchlastywacz) [reply pytripgui/view/qt_view_adapter.py:178]: Done

File: pytripgui/view/qt_view_adapter.py [hunk 5]
```diff
@@ -172,6 +174,14 @@ def checked_items(self):
                 selected.append(widget.data(Qt.UserRole))
         return selected
 
+    def tick_checkboxes(self, items, get_name: Callable):
+        items_to_be_ticked = [
+            list_item for list_item in self._items
+            if get_name(list_item.data(Qt.UserRole)) in [get_name(item) for item in items]
```
Comment (grzanka) [inline pytripgui/view/qt_view_adapter.py:192]: To optimize the code I'd create a dedicated variable first:
```
item_names = [get_name(item) for item in items]
items_to_be_ticked = [
            list_item for list_item in self._items
            if get_name(list_item.data(Qt.UserRole)) in item_names
]
```

File: pytripgui/view/qt_view_adapter.py [hunk 6]
```diff
@@ -172,6 +174,14 @@ def checked_items(self):
...
+        ]
```
Comment (grzanka) [inline pytripgui/view/qt_view_adapter.py:193]: Or better:
```
        for item in items_to_be_ticked:
            if get_name(list_item.data(Qt.UserRole)) in item_names:
                     item.setCheckState(Qt.Checked)
```

Comment (rozchlastywacz) [reply pytripgui/view/qt_view_adapter.py]: Done

File: pytripgui/view/qt_view_adapter.py (hunk 5 above)
Comment (rozchlastywacz) [reply pytripgui/view/qt_view_adapter.py]: Done

File: pytripgui/app_logic/app_callbacks.py [hunk 7]
```diff
@@ -202,7 +202,9 @@ def open_empty_patient_callback(self, patient_item):
             self.app_model.viewcanvases = ViewCanvases()
             self.parent_gui.add_widget(self.app_model.viewcanvases.widget())
 
-        self.app_model.viewcanvases.set_patient(patient)
+        # someone needs to test this, but I think it's unnecessary,
```
Comment (grzanka) [inline pytripgui/app_logic/app_callbacks.py:205]: remove, not necessary

File: pytripgui/canvas_vc/canvas_controller.py [hunk 8]
```diff
@@ -80,19 +85,35 @@ def _update_canvas_view(self):
...
+    def set_patient(self, patient, state: PatientGuiState):
         self._ui.clear()
-        if state is None:
-            state = ProjectionSelector()
-        self._model = PlotModel(state)
+
+        if state:
+            # restore state
+            self._gui_state = state
+            # recreate model with stored positions in each plane
```
Comment (grzanka) [inline pytripgui/canvas_vc/canvas_controller.py:94]: use consistent sequence of actions

File: pytripgui/canvas_vc/canvas_controller.py [hunk 9]
```diff
@@ -80,19 +85,35 @@ def _update_canvas_view(self):
                 self._model.let.prepare_data_to_plot()
                 self._ui.plot_let(self._model.let)
 
-    def set_patient(self, patient, state):
+    def set_patient(self, patient, state: PatientGuiState):
```
Comment (grzanka) [inline pytripgui/canvas_vc/canvas_controller.py:88]: Describe responsibility of this method (when its used)

Comment (grzanka) [reply pytripgui/canvas_vc/canvas_controller.py:88]: in a docstring

**Review by grzanka — CHANGES_REQUESTED** [review]

File: pytripgui/canvas_vc/canvas_controller.py (hunk 8 above)
Comment (rozchlastywacz) [reply pytripgui/canvas_vc/canvas_controller.py:92]: Done

**Review by grzanka — APPROVED** [review]
ok

**Pull request merged**